triton/scripts/export_jit.py

```diff
@@ -0,0 +1,422 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script converts a pretrained transducer model into three jit models:
+# encoder_jit.pt decoder_jit.pt joiner_jit.pt
+"""
+Usage:
+./pruned_transducer_stateless3/export_jit.py \
+  --pretrained-model ./pruned_transducer_stateless3/exp \
+  --output-dir ./pruned_transducer_stateless3/exp \
+  --bpe-model ./bpe.model 
+  
+It will generate three jit files under ouput_dir, then you should put these
+models under corresponding trtion mdoel_repo modules.
+
+"""
+
+import argparse
+import logging
+import os
+
+import sentencepiece as spm
+import torch
+from torch import nn
+
+import torch.nn.functional as F
+from scaling import ScaledConv1d, ScaledEmbedding, ScaledLinear
+
+from conformer_triton import Conformer
+
+class Decoder(nn.Module):
+    """This class modifies the stateless decoder from the following paper:
+
+        RNN-transducer with stateless prediction network
+        https://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=9054419
+
+    It removes the recurrent connection from the decoder, i.e., the prediction
+    network. Different from the above paper, it adds an extra Conv1d
+    right after the embedding layer.
+
+    TODO: Implement https://arxiv.org/pdf/2109.07513.pdf
+    """
+
+    def __init__(
+        self,
+        vocab_size: int,
+        decoder_dim: int,
+        blank_id: int,
+        context_size: int,
+    ):
+        """
+        Args:
+          vocab_size:
+            Number of tokens of the modeling unit including blank.
+          decoder_dim:
+            Dimension of the input embedding, and of the decoder output.
+          blank_id:
+            The ID of the blank symbol.
+          context_size:
+            Number of previous words to use to predict the next word.
+            1 means bigram; 2 means trigram. n means (n+1)-gram.
+        """
+        super().__init__()
+
+        self.embedding = ScaledEmbedding(
+            num_embeddings=vocab_size,
+            embedding_dim=decoder_dim,
+            padding_idx=blank_id,
+        )
+        self.blank_id = blank_id
+
+        assert context_size >= 1, context_size
+        self.context_size = context_size
+        self.vocab_size = vocab_size
+        if context_size > 1:
+            self.conv = ScaledConv1d(
+                in_channels=decoder_dim,
+                out_channels=decoder_dim,
+                kernel_size=context_size,
+                padding=0,
+                groups=decoder_dim,
+                bias=False,
+            )
+
+    def forward(self, y: torch.Tensor) -> torch.Tensor:
```

Comment:
remove need_pad parameters, since it always false when infer, so we don't need to prepare input_need_pad for triton.